module/zfs/spa_errlog.c

```diff
@@ -128,6 +241,300 @@ spa_log_error(spa_t *spa, const zbookmark_phys_t *zb)
+		}
+	}
+
+	/* How many snapshots reference this block. */
+	uint64_t snap_count;
+	error = zap_count(spa->spa_meta_objset,
+	    dsl_dataset_phys(ds)->ds_snapnames_zapobj, &snap_count);
+	if (error != 0) {
+		dsl_dataset_rele(ds, FTAG);
+		return (error);
+	}
+
+	if (snap_count == 0) {
+		/* File system has no snapshot. */
+		dsl_dataset_rele(ds, FTAG);
+		return (0);
+	}
+
+	uint64_t *snap_obj_array = kmem_alloc(snap_count * sizeof (uint64_t),
+	    KM_SLEEP);
+
+	int aff_snap_count = 0;
+	uint64_t snap_obj = dsl_dataset_phys(ds)->ds_prev_snap_obj;
+	uint64_t snap_obj_txg = dsl_dataset_phys(ds)->ds_prev_snap_txg;
+	uint64_t zap_clone = dsl_dir_phys(ds->ds_dir)->dd_clones;
+
+	/* Check only snapshots created from this file system. */
+	while (snap_obj != 0 && zep->zb_birth < snap_obj_txg &&
+	    snap_obj_txg <= txg_to_consider) {
+
+		dsl_dataset_rele(ds, FTAG);
+		error = dsl_dataset_hold_obj(dp, snap_obj, FTAG, &ds);
+		if (error != 0)
+			goto out;
+
+		if (dsl_dir_phys(ds->ds_dir)->dd_head_dataset_obj != fs)
+			break;
+
+		boolean_t affected = B_TRUE;
+		if (check_snapshot) {
+			uint64_t blk_txg;
+			error = find_block_txg(ds, zep, &blk_txg);
+			affected = (error == 0 && zep->zb_birth == blk_txg);
+		}
+
+		if (affected) {
+			snap_obj_array[aff_snap_count] = snap_obj;
+			aff_snap_count++;
+
+			if (!only_count) {
+				zbookmark_phys_t zb;
+				zep_to_zb(snap_obj, zep, &zb);
+				if (copyout(&zb, (char *)addr + (*count - 1) *
+				    sizeof (zbookmark_phys_t),
+				    sizeof (zbookmark_phys_t)) != 0) {
+					dsl_dataset_rele(ds, FTAG);
+					error = SET_ERROR(EFAULT);
+					goto out;
+				}
+				(*count)--;
+			} else {
+				(*count)++;
+			}
+		}
+		snap_obj_txg = dsl_dataset_phys(ds)->ds_prev_snap_txg;
+		snap_obj = dsl_dataset_phys(ds)->ds_prev_snap_obj;
+	}
+	dsl_dataset_rele(ds, FTAG);
+
+	/* Check clones referencing the error block. */
+	if (zap_clone != 0 && aff_snap_count > 0) {
+		zap_cursor_t zc;
+		zap_attribute_t za;
+
+		for (zap_cursor_init(&zc, spa->spa_meta_objset, zap_clone);
+		    zap_cursor_retrieve(&zc, &za) == 0;
+		    zap_cursor_advance(&zc)) {
+
+			dsl_pool_t *dp = spa->spa_dsl_pool;
+			dsl_dataset_t *clone;
+			error = dsl_dataset_hold_obj(dp, za.za_first_integer,
+			    FTAG, &clone);
+
+			if (error != 0) {
+				zap_cursor_fini(&zc);
+				goto out;
+			}
+
+			boolean_t found = B_FALSE;
+			for (int i = 0; i < aff_snap_count && !found; i++) {
+				found =
+				    (dsl_dir_phys(clone->ds_dir)->dd_origin_obj
+				    == snap_obj_array[i]);
+			}
+			dsl_dataset_rele(clone, FTAG);
+
+			if (!found)
+				continue;
+
+			error = check_filesystem(spa, za.za_first_integer, zep,
+			    count, addr, only_count);
+
+			if (error != 0) {
+				zap_cursor_fini(&zc);
+				goto out;
+			}
+		}
+		zap_cursor_fini(&zc);
+	}
+out:
+	kmem_free(snap_obj_array, sizeof (*snap_obj_array));
+	return (error);
+}
+
+static int
+find_top_affected_fs(spa_t *spa, uint64_t head_ds, zbookmark_err_phys_t *zep,
```

Comment:
I will add the comments, and move this to `dsl_dataset.c`. However, I think this function returns the `dd_head_dataset_obj` of the oldest snapshot which I believe is different than the oldest snapshot. Please correct me if I am wrong.

Reply:
I see.  It might make sense to have a simpler, more general function like I described above, and then in spa_errlog.c we can go from the oldest snapshot obj to its head dataset (even though this does involve one more call to dsl_dataset_hold_obj()).